linux/install-dkms

```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+#
+# Usage: ./install-dkms [640 480]
+
+set -e
+REPO_PATH="$(/usr/bin/dirname $([ -L $0 ] && /bin/readlink -f $0 || echo $0))"
+source "${REPO_PATH}/install_shell_func_and_conf"
+
+
+check_params "$1" "$2"
+
+# Installing droidcam binaries. Requires make to have succeeded.
+if [ -f "${REPO_PATH}/droidcam" ]; then 
+    echo "Installing droidcam in /usr/bin" 
+    cp "${REPO_PATH}/droidcam" /usr/bin/
+else
+    echo "droidcam was not previously built. Installation skip."
+fi
+if [ -f "${REPO_PATH}/droidcam-cli" ]; then 
+    echo "Installing droidcam-cli in /usr/bin" 
+    cp "${REPO_PATH}/droidcam-cli" /usr/bin/
+else
+    echo "droidcam-cli was not previously built. Installation skip."
+fi
+
+# Now setting up the module as a DKMS module
+
+# Creating link with correct name in /usr/src to the v4l2loopback directory
+cd "${REPO_PATH}/${V4L2_LOOPBACK_DIR}"
+if [ -d "/usr/src/${V4L2_LOOPBACK_DC}-${V4L2_LOOPBACK_VERSION}" ] || [ -L "/usr/src/${V4L2_LOOPBACK_DC}-${V4L2_LOOPBACK_VERSION}" ]; then
+    # Skipping link creation...
+    echo "'usr/src/${V4L2_LOOPBACK_DC}-${V4L2_LOOPBACK_VERSION}' already exists. Skipping creation."
+else
+    ln -s "`pwd`" "/usr/src/${V4L2_LOOPBACK_DC}-${V4L2_LOOPBACK_VERSION}"
```

Comment:
The install scripts are meant to go into the final portable install package, which users download to /tmp right now.
Rather than linking we'd ned to copy/move the drivers files into /usr/src

Reply:
Humm, portable install package ? What do you mean exactly by that ?
I don't really know such a mechanism under Linux (of course you have Snaps, AppImages, docker/lxc containers etc... but I doubt this is what you mean).
I just know such a mechanism under Windows but I don't really see the point. Could you please tell me more about what you think there ?

Nevertheless, two remarks:

* If we talk here about some kind of app containment, isn't the fact that you copy a kernel module into the kernel tree somehow breaking that rule anyway ?
* More specifically about DKMS and `/usr/src`. DKMS uses the content of what is in `/usr/src` to actually **copy** it elsewhere to keep track of it (in `/var/lib/dkms/module_name/module_version`). There it keeps source, tar balls created, deb files, builds and logs. Thus about that link in `/usr/src`, I was even wondering if I should not actually remove it after the install...

Reply:
I noticed that after the original install you actually copy the uninstall script in `/opt`. I could do the same after the DKMS install, but I was wondering why you actually do that.
Anyways, if you want to do that, do you think I should copy it to `/opt/droidcam-uninstall-dkms` or `/opt/droidcam-uninstall` like the original (that would sound cleaner to me) ?

Reply:
>     * More specifically about DKMS and `/usr/src`. DKMS uses the content of what is in `/usr/src` to actually **copy** it elsewhere to keep track of it (in `/var/lib/dkms/module_name/module_version`). There it keeps sources, tar balls created, deb files, builds and logs. Thus about that link in `/usr/src`, I was even wondering if I should not actually remove it after the install... But I just kept it because personally I prefer a harmless broken link that reminds me something was there, than nothing and forgetting. But that's personal. Tell me where your preference goes. Copying there would introduce a lot of side problems (is it in sync ?  Therefore should we blast it before a new install ? etc...)

Actually I was wrong. DKMS does a symlink from `/usr/src/...` to `/var/lib/dkms/module_name/module_version/source`...
Thus implies probably a copy after all. Then I would clearly have preferred a separated git repo for the module :(

Reply:
Ok last push should have fixed/enhanced everything.

*  `v4l2loopback_dc` module is now copied to `/usr/src` instead of just symlinked.
* The `uninstall-dkms` script is no more relying on `install.common` and therefore is copied to `/opt/droidcam-uninstall` during install.
* The `dkms.conf` is no more in the repository but instead generated in `/usr/src` during install process, thus module version in `dkms.conf` will remain consistent with the one declared in `install.common`.

I hope this solves all your potential concerns.

Reply:
By portable package I was referring to the download zip package at dev47apps.com.